### PR TITLE
Legacy widget's preview functionality is broken when the page is moved

### LINKED ISSUE #30049: Widgets preview uses a relative URL that breaks when the page is moved
## Description

The experimental widgets UI has a preview function, **however this preview uses a relative URL**.

Normally this is fine, however, I've been trying to add this to frontenberg, which means the code is no longer running from the `/wp-admin/` URL. As a result all previews are now Nginx 404 error pages on my local environment.

## Expected behaviour

A full URL, e.g.

```
<iframe src="https://one.wordpress.test/wp-admin/themes.php?page=gutenberg-widgets&amp;widget-preview%5BwidgetClass%5D=WP_Widget_Recent_Comments&amp;widget-preview%5Bname%5D=Recent%20Comments&amp;widget-preview%5BidBase%5D=recent-comments&amp;widget-preview%5Bnumber%5D=2&amp;widget-preview%5Binstance%5D%5Btitle%5D=&amp;widget-preview%5Binstance%5D%5Bnumber%5D=5&amp;widget-preview%5B__internalWidgetId%5D=recent-comments-2&amp;widget-preview%5BsidebarId%5D=wp_inactive_widgets" height="138" class="wp-block-legacy-widget__preview"></iframe>
```

## Actual behaviour

`src="themes.php?....` is used, resulting in this:

Screenshot 2021-03-19 at 22 55 00

## WordPress information
- WordPress version: 5.7.0
- Gutenberg version: 10.2.0
- Are all plugins except Gutenberg deactivated? Yes
- Are you using a default theme (e.g. Twenty Twenty-One)? No ( modified Frontenberg )

## Device information
- Device: MBP 13" Intel
- Operating system: MacOS 11.2
- Browser: Chrome 89

### PR DESCRIPTION
**Pull request by anton-vlasenko**
## Description
Legacy Widget block has a preview function. However, it uses a relative URL.
`iframe` cannot load the widgets.php file when the code is no longer running from the /wp-admin/ URL.
This PR fixes that by using a REST API endpoint instead of the widgets.php file to fetch preview HTML. It doesn't depend on the admin URL.
Fixes #30049 

Based on #34230

## How has this been tested?
1. Go to Appearance -> Widgets page.
2. Click on the "plus sign" button at the very bottom of the page.
3. Add a `Legacy Widget` block.
4. Select `Meta` widget from the dropdown.
5. Click somewhere on the background of the page (not on the `Meta` widget itself). You should see a preview of the `Meta` widget now.
6. Open your browser's developer tools and find an `<iframe />` element with the `wp-block-legacy-widget__edit-preview-iframe`  class. Check its `srcdoc` attribute.

#### Expected result:
The `srcdoc` attribute of the `<iframe />` should contain html code of the preview.
Please see the screenshots section for reference.

## Screenshots 
Screenshot 2021-08-12 at 20 00 39
Screenshot 2021-08-19 at 21 59 05

## Types of changes
Bug fix.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (kevin940726) [inline packages/widgets/src/blocks/legacy-widget/edit/preview.js]: It's probably better to set state here instead of manipulating DOM directly. We'll have to create a new state and call the setState function here.

```js
const [ srcDoc, setSrcDoc ] = useState( '' );

renderPreview( idBase, instance ).then( ( response ) => {
  setSrcDoc( response.preview );
} );

<iframe srcDoc={ srcDoc } />
```

We'll probably also want to add some error handling and abort the fetch when the component is unmounted too. Which can be a follow-up PR.

Comment (kevin940726) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: Also, this request isn't depend on the reference of the iframe node itself, so I don' think it's necessary to put it in `useRefEffect`. Instead, using `useEffect` should be enough.

We should also drop the custom `previewLoaded` field too.

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (kevin940726) [inline packages/widgets/src/blocks/legacy-widget/edit/preview.js]: Nit: I'd probably go for `getPreviewHTML` or something like that but it's just my preference.

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (adamziel) [inline packages/widgets/src/blocks/legacy-widget/edit/preview.js:131]: No definite recommendations, but I wonder if we still need this `onLoad` handler if the loading is now done via api fetch.

Comment (kevin940726) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js:131]: `apiFetch` only fetches the html string, this `onLoad` handler is needed when the actual iframe content is loaded by the browser.

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: > We should also drop the custom previewLoaded field too.

I don't think we need to drop `previewLoaded` entirely as we only need to fetch the preview HTML once.
But I agree that we probably need to create a new state for that.
We need some variable that we can use as a flag to control if the `apiFetch` function has already been called. We can't use `srcdoc` for that because if the apiFetch request fails, the srcdoc attribute will be empty and we can't use it to find out whether we called apiFetch or not.

Comment (kevin940726) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: Yep, I mean to `useState` for that rather than adding unsupported properties to DOM elements.

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: Fixed in 3ea1587d917b7155a477762df9f820298b5676dc

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: @kevin940726 Thank you for the feedback!
I hope I've fixed all the issues you mentioned.
Could you please check it?

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (kevin940726) [inline packages/widgets/src/blocks/legacy-widget/edit/preview.js]: These should all be in `useEffect` so that we can avoid creating an abortController and a function for every render.

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (kevin940726) [inline packages/widgets/src/blocks/legacy-widget/edit/preview.js]: I don't think we need to add custom error. We can just re-throw the error here.

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (kevin940726) [inline packages/widgets/src/blocks/legacy-widget/edit/preview.js]: Can be just be:

```suggestion
			abortController?.abort();
```

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (kevin940726) [inline packages/widgets/src/blocks/legacy-widget/edit/preview.js]: We'll want to re-run this effect whenever `idBase` or `instance` changes.

```suggestion
	}, [ idBase, instance ] );
```

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (kevin940726) [inline packages/widgets/src/blocks/legacy-widget/edit/preview.js]: We don't need this state, it should already be handled by the effect, and we'll also want to re-run the effect anyway. `srcDoc` state seems to be enough.

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (kevin940726) [inline packages/widgets/src/blocks/legacy-widget/edit/preview.js:58]: Should probably revert the change since that I think the diff is the same?

**Review by kevin940726 — CHANGES_REQUESTED** [review]

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: Yes, we don't have to add a custom error, but it will not hurt if we add it.  That custom error gives user some additional context, e.g., why the error happened. I mean, it's clear from the error message that it happened during the fetching of the preview. IMO If we just throw an error it will not be that clear.

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: Good catch.
Fixed in 982c2053d656220b0047641133763d211b6d6868

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: Fixed in 982c2053d656220b0047641133763d211b6d6868

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: Good catch!
Fixed in 982c2053d656220b0047641133763d211b6d6868

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js:58]: The main idea behind this was to apply so called `return early` approach (it's just a refactoring). We don't need to include all that code into parenthesis to avoid nesting.
Please see an example [here](https://github.com/jupeter/clean-code-php/blob/master/README.md#avoid-nesting-too-deeply-and-return-early-part-1). That example is for PHP but it also can be applied to JavaScript.

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: Good catch.
Fixed in 982c2053d656220b0047641133763d211b6d6868

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (kevin940726) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: It does increase the bundle size though, and the error message can't be minified. We could potentially use `@wordpress/warning` for this, it does some magic to exclude the error from prod.
I don't think throwing an error is that bad though, in fact, I think it's often time better than a simple `console.error` because it has stack traces available. And it's also a lot more visible so that we can monitor the unexpected early.

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (kevin940726) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js:58]: Yeah, I got it 👍. What I mean here is mostly about unrelated changes in this PR though. This change is not related to this PR and is increasing the diff to make it harder to review.

I don't have a strong preference about the coding style. I think it's fine when it's only one level nested, but doing early return also looks good to me.

Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js:58]: Yes, it increases the diff size.
On the other hand, I feel like creating a separate PR for it would be overkill. We will have to create a different issue for it as well.

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: > It does increase the bundle size though, and the error message can't be minified. 

Let's not use it then :) 
Fixed in a37caf465709f29a85b52fdd023e62681298712b

File: packages/widgets/src/blocks/legacy-widget/edit/preview.js
Comment (kevin940726) [inline packages/widgets/src/blocks/legacy-widget/edit/preview.js]: I believe we can remove this comment now?

Comment (anton-vlasenko) [reply packages/widgets/src/blocks/legacy-widget/edit/preview.js]: Fixed in 98d0fafc27ee1bb29952446a6bcc394b7c510ec1

**Comment by anton-vlasenko** [comment]
I've merged #34230 into this PR to make the e2e tests pass.
~~So please ignore the code from  #34230 during the code review.~~
UPD: now it should only show changes related to this PR because I've changed the target branch.

**Comment by kevin940726** [comment]
@anton-vlasenko You can target the base branch to #34230's branch so that it won't show the commits from that PR. Just have to remember to rebase to trunk once it's merged :).

**Comment by anton-vlasenko** [comment]
@kevin940726 
> @anton-vlasenko You can target the base branch to #34230's branch so that it won't show the commits from that PR. Just have to remember to rebase to trunk once it's merged :).

That's a good idea. I tried to do it as you said and I got this popup message:
Screenshot 2021-08-25 at 11 54 32
I understand it's doable but I'm afraid to screw something up.

**Comment by kevin940726** [comment]
@anton-vlasenko I believe you just have to remove the commits that's already on that branch and rebase your branch onto that.

**Comment by anton-vlasenko** [comment]
> @anton-vlasenko I believe you just have to remove the commits that's already on that branch and rebase your branch onto that.

@kevin940726 Never mind.
I've changed the target branch of this PR (just as you suggested in [this comment](url)).
> I believe you just have to remove the commits that's already on that branch and rebase your branch onto that.

I'm sorry, but I don't understand why I should remove commits and rebase onto #34230 if we already have the commits from #34230 merged into this branch. The only difference would be in the git history log ( we will have separate commits in case of rebase vs. a single commit in case of a merge).

File: packages/e2e-tests/specs/widgets/editing-widgets.test.js
Comment (kevin940726) [inline packages/e2e-tests/specs/widgets/editing-widgets.test.js:636]: Why do we need to change this?

Comment (anton-vlasenko) [reply packages/e2e-tests/specs/widgets/editing-widgets.test.js:636]: > @anton-vlasenko I believe you just have to remove the commits that's already on that branch and rebase your branch onto that.

Done.

Comment (anton-vlasenko) [reply packages/e2e-tests/specs/widgets/editing-widgets.test.js:636]: I've changed this because otherwise e2e tests fail.
We cannot "detect"  the needed frame by its URL because we've removed the `src` attribute.
So now we are selecting the related iframe element by its class name. Then we get the needed frame using the iframe element mentioned above.

Comment (kevin940726) [reply packages/e2e-tests/specs/widgets/editing-widgets.test.js:636]: Oh cool, yeah!

We've experienced some flaky situations with iframe stuffs though so it'd be cool if we can re-run this several times before we merge this 👍.

**Review by kevin940726 — APPROVED** [review]
Code LGTM 👍. We need to wait for #34230 before merging this though, or just merge this onto that branch.

**Comment by anton-vlasenko** [comment]
This PR has been closed by GitHub (unintentionally).
So let's continue with the review [here](https://github.com/WordPress/gutenberg/pull/34384).

**Pull request closed without merging**
